pkg/scalers/metrics_api_scaler.go

```diff
@@ -13,20 +13,53 @@ import (
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/metrics/pkg/apis/external_metrics"
 	"net/http"
+	neturl "net/url"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"strconv"
+	"strings"
+	"time"
 )
 
 type metricsAPIScaler struct {
 	metadata *metricsAPIScalerMetadata
+	client   *http.Client
 }
 
 type metricsAPIScalerMetadata struct {
 	targetValue   int
 	url           string
 	valueLocation string
+
+	//apiKeyAuth
+	enableAPIKeyAuth bool
+	method           string // +default is header
+	// keyParamName  is either header key or query param used for passing apikey
+	// default header is "X-API-KEY", defaul query param is "api_key"
+	keyParamName string
+	apiKey       string
+
+	//base auth
+	enableBaseAuth bool
+	username       string
+	password       string // +optional
+
+	//client certification
+	enableTLS bool
+	cert      string
+	key       string
+	ca        string
 }
 
+const defaultTimeOut = 3 * time.Second
+
+type authenticationType string
+
+const (
+	apiKeyAuth authenticationType = "apiKeyAuth"
+	basicAuth                     = "basicAuth"
+	tlsAuth                       = "tlsAuth"
```

Comment:
I'm not 100% but I think we should add `authenticationType` to two other methods, otherwise go will resolve the type from assigned value?

Reply:
oh didn't know that automatic type assignment works for zero indexed type declaration i.e. with iota. In testing, this didn't get caught because of its string representation.